lang/thread.md

```diff
@@ -0,0 +1,163 @@
+# MiniRust Thread
+
+This defines the state that makes up a thread of execution in MiniRust. It is used by the Abstract Machine to make progress on.
+
+```rust
+// This type contains all information the abstract machine needs
+// to continue a thread of execution.
+pub struct Thread<M: Memory> {
+    state: ThreadState,
+
+    // The stack.
+    stack: List<StackFrame<M>>,
+}
+
+pub enum ThreadState {
+    // The thread is enabled and can get executed.
+    Enabled,
+    // The thread is trying to join another thread and is blocked until that thread finishes.
+    BlockedOnJoin(ThreadId),
+    // The thread has terminated. We do not delete terminated threads until they are joined.
+    Terminated,
+}
+
+type ThreadId = Int;
+```
+
+Next, we define how to create a thread.
+
+```rust
+impl<M: Memory> Thread<M> {
+    fn new(func: Function) -> Self {
+        // Setup the initial stack frame.
+        // For the main thread, well-formedness ensures that the func has
+        // no return value and no arguments.
+        // For any other threads, the spawn intrinsic ensures
+        // that the func has no arguments.
+        let init_frame = StackFrame {
+            func,
+            locals: Map::new(),
+            caller_return_info: None,
+            next_block: func.start,
+            next_stmt: Int::ZERO,
+        };
+
+        Self {
+            state: ThreadState::Enabled,
+            stack: list![init_frame],
+        }
+    }
+}
+```
+
+Some functionality.
+
+```rust
+impl<M: Memory> Thread<M> {
+    fn cur_frame(&self) -> StackFrame<M> {
+        self.stack.last().unwrap()
+    }
+
+    fn mutate_cur_frame<O>(&mut self, f: impl FnOnce(&mut StackFrame<M>) -> O) -> O {
+        if self.stack.is_empty() {
+            panic!("`mutate_cur_frame` called on empty stack!");
+        }
+
+        let last_idx = self.stack.len() - 1;
+        self.stack.mutate_at(last_idx, f)
+    }
+
+    fn mutate_stack<O>(&mut self, f: impl FnOnce(&mut List<StackFrame<M>>) -> O) -> O {
+        f(&mut self.stack)
+    }
```

Comment:
This doesn't look like we need a `mutate` function? We need one when mutating things inside a GC ptr but here that is not the case.